srai/embedders/geovex/dataset.py

```diff
@@ -0,0 +1,196 @@
+                            (data.index.get_loc(_h), get_local_ij_index(h3_index, _h))
+                            for _h in neighbors
+                        ],
+                    )
+                )
+            else:
+                # some of the neighbors are not in the dataset, add the h3_index to the invalid h3s
+                invalid_h3s.add(h3_index)
+        return invalid_h3s, valid_h3s
+
+    def __len__(self) -> int:
+        """
+        Returns the number of valid h3 indices in the dataset.
+
+        Returns:
+            int: Number of valid h3 indices in the dataset.
+        """
+        return len(self._valid_cells)
+
+    def __getitem__(self, index: Any) -> "torch.Tensor":
+        """
+        Return a single item from the dataset.
+
+        Args:
+            index (Any): The index of dataset item to return
+
+        Returns:
+            HexagonalDatasetItem: The dataset item
+        """
+        _, target_idx, neighbors_idxs = self._valid_cells[index]
+        return self._build_tensor(target_idx, neighbors_idxs)
+
+    def _build_tensor(
+        self, target_idx: int, neighbors_idxs: List[Tuple[int, Tuple[int, int]]]
+    ) -> "torch.Tensor":
+        import torch
+
+        # build the 3d tensor
+        # it is a tensor with diagonals of length neighbor_k_ring
+        # the diagonals are the neighbors of the target h3
+        # the target h3 is in the center of the tensor
+        # the tensor is 2*neighbor_k_ring + 1 x 2*neighbor_k_ring + 1 x 2*neighbor_k_ring + 1
+        # make a tensor of zeros, padded by 1 zero all around to make it even for the convolutions
+        tensor: torch.Tensor = torch.zeros(
+            (
+                self._N,
+                2 * self._k + 2,
+                2 * self._k + 2,
+            )
+        )
+
+        # set the target h3 to the center of the tensor
+        tensor[
+            :,
+            self._k,
+            self._k,
+        ] = self._data_torch[target_idx]
+
+        # set the neighbors of the target h3 to the diagonals of the tensor
+        for neighbor_idx, (i, j) in neighbors_idxs:
+            tensor[:, self._k + i, self._k - j] = self._data_torch[neighbor_idx]
+
+        # return the tensor and the target (which is same as the tensor)
+        # should we return the target as a copy of the tensor?
+        return tensor
+
+    def _assert_k_ring_correct(self, k_ring: int) -> None:
+        if k_ring < 2:
+            raise ValueError(f"k_ring must be at least 2, but was {k_ring}")
+
+    def _assert_h3_neighbourhood(self, neighbourhood: H3Neighbourhood) -> None:
+        # force that the neighbourhood is an H3Neighbourhood,
+        # because we need the get_ij_index method
+        if not isinstance(neighbourhood, H3Neighbourhood):
+            raise ValueError(
+                f"neighbourhood has to be an H3Neighbourhood, but was {type(neighbourhood)}"
+            )
+
+    def get_ordered_index(self) -> List[str]:
```

Comment:
Should it be named `get_ordered_index` or `get_valid_cells` (looking at the function below)? I see that there is no sorting / ordering in this function.

Reply:
It is sorted per-say, as it maintains the order of the input count Dataframe, but I am okay with the rename